lib/index.js

```diff
@@ -52,14 +53,60 @@ function compileBody(str, options){
   var debug_sources = {};
   debug_sources[options.filename] = str;
   var dependencies = [];
+  var plugins = options.plugins || [];
+  plugins = plugins.map(function (plugin) {
+    if (typeof plugin === 'string') {
+      var opts = {
+        basedir: path.dirname(module.parent.filename)
+      };
+      // Test if plugin name is local (i.e. `./my-plugin`)
+      // Regex from https://github.com/substack/node-resolve/blob/5cae82fb22cb64d5b72f703c787dc0fd418ed412/lib/sync.js#L21
+      if (/^(?:\.\.?(?:\/|$)|\/|([A-Za-z]:)?[\\\/])/.test(plugin)) {
+        return require(resolve(plugin, opts))();
+      } else {
+        return require(resolve('jade-plugin-' + plugin, opts))();
+      }
+    }
+    return plugin;
+  });
+
   try {
     var ast = load.string(str, options.filename, {
-      lex: lex,
+      lex: function (str, filename) {
+        getPlugin('preLex').forEach(function (func) {
+          str = func(str);
+        });
+        var tokens = lex(str, filename, {
+          plugins: getPlugin('lex')
+        });
+        getPlugin('postLex').forEach(function (func) {
+          tokens = func(tokens);
+        });
+        return tokens;
+      },
       parse: function (tokens, filename) {
+        getPlugin('preParse').forEach(function (func) {
+          tokens = func(tokens);
+        });
         tokens = stripComments(tokens, { filename: filename });
-        return parse(tokens, filename);
+        var ast = parse(tokens, filename, {
+          plugins: getPlugin('parse')
+        });
+        getPlugin('postParse').forEach(function (func) {
+          ast = func(ast);
+        });
+
+        getPlugin('preLoad').forEach(function (func) {
+          ast = func(ast);
+        });
+        return ast;
       },
       resolve: function (filename, source) {
+        var plugins = getPlugin('loadResolve');
+        for (var i = 0; i < plugins.length; i++) {
+          var result = plugins[i](filename, source, options);
+          if (result) return result;
```

Comment:
The empty string should be considered a valid file path here.  Lets do `result !== undefined` as our test.